src/cljs/proton/layers/lang/haskell/core.cljs

```diff
@@ -0,0 +1,85 @@
+
+(defmethod init-layer! :lang/haskell
+  [_ {:keys [config layers]}]
+  (helpers/console! "init" :lang/haskell)
+  (register-layer-dependencies :tools/linter [:linter-hlint]))
+
+(defmethod get-keybindings :lang/haskell []
+  {:f {:category "files"
+       := {:action "ide-haskell:prettify-file"
+           :title "format/prettify code"}
+       :c {:action "haskell-ghc-mod:check-file"
+           :title "check file"}
+       :l {:action "haskell-ghc-mod:lint-file"
+           :title "lint file"}}}
+  {:e {:category "errors"
+       :n {:action "ide-haskell:next-error"}
+       :p {:action "ide-haskell:prev-error"}}})
+
+(defmethod init-package [:lang/haskell :ide-haskell] []
+  (helpers/console! "init ide-haskell package" :lang/haskell)
+  (mode/define-package-mode :ide-haskell
+    {:mode-keybindings
+      {:t {:category "toggles"
+           :o {:action "ide-haskell:toggle-output" :title "ide-haskell output"}}}})
+  (mode/link-modes :haskell-major-mode (mode/package-mode-name :ide-haskell)))
+
+(defmethod init-package [:lang/haskell :haskell-ghc-mod] []
+  (helpers/console! "init haskell-ghc-mod package" :lang/haskell)
+  (mode/define-package-mode :haskell-ghc-mod
+    {:mode-keybindings
+      {:h {:category "help/docs"
+           :t {:action "haskell-ghc-mod:show-type" :title "type"}
+           :i {:action "haskell-ghc-mod:show-info" :title "info"}}
+       :j {:category "jump"
+           :d {:action "haskell-ghc-mod:go-to-declaration" :title "declaration"}}
+       :i {:category "insert"
+           :t {:action "haskell-ghc-mod:insert-type" :title "type"}
+           :i {:action "haskell-ghc-mod:insert-import" :title "import"}}}})
+  (mode/link-modes :haskell-major-mode (mode/package-mode-name :haskell-ghc-mod)))
+
+(defmethod init-package [:lang/haskell :ide-haskell-cabal] []
+  (helpers/console! "init ide-haskell-cabal package" :lang/haskell)
+  (mode/define-package-mode :ide-haskell-cabal
+    {:mode-keybindings
+      {:c {:category "compiler"
+           :s {:action "ide-haskell-cabal:set-build-target" :title "select target"}
+           :b {:action "ide-haskell-cabal:build" :title "build"}
+           :c {:action "ide-haskell-cabal:clean" :title "clean"}
+           :t {:action "ide-haskell-cabal:test" :title "test"}}}})
+  (mode/link-modes :haskell-major-mode (mode/package-mode-name :ide-haskell-cabal)))
+
+(defmethod init-package [:lang/haskell :haskell-hoogle] []
+  (helpers/console! "init haskell-hoogle package" :lang/haskell)
+  (mode/define-package-mode :haskell-hoogle
+    {:mode-keybindings
+      {:h {:category "help/docs"
+           :l {:action "haskell-hoogle:lookup" :title "hoogle search"}}}})
+  (mode/link-modes :haskell-major-mode (mode/package-mode-name :haskell-hoogle)))
+
+(defmethod get-packages :lang/haskell []
+ [:autocomplete-haskell
+   :haskell-ghc-mod
+   :haskell-hoogle
+   :ide-haskell
+   :ide-haskell-cabal
+   :language-haskell])
+
+(defmethod describe-mode :lang/haskell []
```

Comment:
Need help here. Do I need to add cabal as a grammar and then add the `.cabal` extension? Auto-indent on cabal isn't working as I would expect and I'm not if it's my fault or downward dependency.